### PR TITLE
Update to pydantic v2 in apps/scout

### PR DESCRIPTION
**Pull request by islean**
## Description

Updates to Pydantic v2 in apps/scout and moves any validators to its file to keep the models as clean as possible. This includes usage of model_validate and annotating any validators instead of keeping them at the bottom of the base model.

### Added

- Package for keeping validators for base models.

### Fixed

- Pydantic v2 now used. 

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b update-pydantic-scout -a
    ```

### How to test

- [x] Run `cg upload upload-case-to-scout <case_id>`.

### Expected test outcome

- [x] Check that the models are constructed without crashing.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by islean** [comment]
Any input on whether I should create a directory for validators.py to be able to standardise across the directories or keep it as is, since it is one file only in this case.

File: cg/apps/scout/validators.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from cg.constants.subject import Gender, PlinkGender, RelationshipStatus
+
+
+def set_parent_if_missing(parent: str):
```
Comment (henrikstranneheim) [inline cg/apps/scout/validators.py:4]: Add return hint

File: cg/apps/scout/validators.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+    return RelationshipStatus.HAS_NO_PARENT if parent is None else parent
+
+
+def set_gender_if_other(gender: str):
```
Comment (henrikstranneheim) [inline cg/apps/scout/validators.py:8]: Add return type hint

**Review by henrikstranneheim — APPROVED** [review]
Nice, are there unit tests for each validator? Would be great to add if there are none. 👍

**Review by diitaz93 — APPROVED** [review]
Looks good 👍

**Comment by diitaz93** [comment]
> Any input on whether I should create a directory for validators.py to be able to standardise across the directories or keep it as is, since it is one file only in this case.

I think the way you implement it is excellent! But I will double-check what is suggested by the documentation.

**Comment by islean** [comment]
> Nice, are there unit tests for each validator? Would be great to add if there are none. 👍

Added now!

File: tests/apps/scout/test_scout_models.py [hunk 3]
```diff
@@ -84,7 +89,56 @@ def test_convert_other_sex(other_sex_case_output: str):
...
     # THEN assert that the sex has been converted to "0"
-    assert case_obj.dict()["individuals"][0][Pedigree.SEX] == PlinkGender.UNKNOWN
+    assert case_obj.model_dump()["individuals"][0][Pedigree.SEX] == PlinkGender.UNKNOWN
+
+
+def test_set_parent_when_provided():
+    """Test to validate that the parent value is not altered if a string is provided."""
+
+    # GIVEN a valid parent
+    father: str = "father"
```
Comment (henrikstranneheim) [inline tests/apps/scout/test_scout_models.py:102]: There are constants for these pedigree terms

File: tests/apps/scout/test_scout_models.py [hunk 4]
```diff
@@ -84,7 +89,56 @@ def test_convert_other_sex(other_sex_case_output: str):
...
+    # THEN the returned string should have been set to RelationshipStatus.HAS_NO_PARENT
+    assert validated_parent == RelationshipStatus.HAS_NO_PARENT
+
+
+def test_set_gender_if_provided():
+    """Test to validate that the gender is not altered when set and not Gender.OTHER."""
+
+    # GIVEN a gender which is not Gender.OTHER as input
+    gender: PlinkGender = PlinkGender.FEMALE
+    # WHEN running "set_gender_if_other"
```
Comment (henrikstranneheim) [inline tests/apps/scout/test_scout_models.py:128]:
```suggestion

    # WHEN running "set_gender_if_other"
```

File: tests/apps/scout/test_scout_models.py [hunk 5]
```diff
@@ -84,7 +89,56 @@ def test_convert_other_sex(other_sex_case_output: str):
...
+    # THEN the returned string should not have been altered
+    assert validated_gender == gender
+
+
+def test_set_gender_if_other():
+    """Test to validate that the gender is altered when set to Gender.OTHER."""
+
+    # GIVEN Gender.OTHER as input
+    gender: Gender = Gender.OTHER
+    # WHEN running "set_gender_if_other"
```
Comment (henrikstranneheim) [inline tests/apps/scout/test_scout_models.py:140]:
```suggestion

    # WHEN running "set_gender_if_other"
```

**Review by henrikstranneheim — APPROVED** [review]
👍

File: tests/apps/scout/test_scout_models.py (hunk 3 above)
Comment (islean) [reply tests/apps/scout/test_scout_models.py]: Fixed!

**Comment by islean** [comment]
Test not successful:

```shell
θ69° [isak.ohlsson@hasta:~] [S_base] 5s $ cg upload --help
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_main/bin/cg", line 5, in <module>
    from cg.cli.base import base
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/base.py", line 11, in <module>
    from cg.cli.delete.base import delete
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/delete/base.py", line 6, in <module>
    from cg.cli.delete.case import delete_case
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/delete/case.py", line 8, in <module>
    from cg.cli.get import get_case as print_case
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/get.py", line 6, in <module>
    from cg.models.cg_config import CGConfig
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/models/cg_config.py", line 16, in <module>
    from cg.apps.scout.scoutapi import ScoutAPI
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/apps/scout/scoutapi.py", line 8, in <module>
    from cg.apps.scout.scout_export import ScoutExportCase, Variant
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/apps/scout/scout_export.py", line 51, in <module>
    class ScoutExportCase(BaseModel):
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_model_construction.py", line 179, in __new__
    types_namespace=types_namespace,
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_model_construction.py", line 431, in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/main.py", line 533, in __get_pydantic_core_schema__
    return __handler(__source)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_schema_generation_shared.py", line 82, in __call__
    schema = self._handler(__source_type)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 281, in generate_schema
    obj, from_dunder_get_core_schema=from_dunder_get_core_schema, from_prepare_args=from_prepare_args
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 301, in _generate_schema_for_type
    schema = self._generate_schema(obj)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 519, in _generate_schema
    return self._model_schema(obj)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 370, in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 370, in <dictcomp>
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 674, in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 716, in _common_field_schema
    annotations,
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 1405, in _apply_annotations
    schema = get_inner_schema(source_type)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_schema_generation_shared.py", line 82, in __call__
    schema = self._handler(__source_type)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_generate_schema.py", line 1487, in new_handler
    schema = metadata_get_schema(source, get_inner_schema)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/functional_validators.py", line 123, in __get_pydantic_core_schema__
    info_arg = _inspect_validator(self.func, 'before')
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/pydantic/_internal/_decorators.py", line 507, in inspect_validator
    sig = signature(validator)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/inspect.py", line 3075, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/inspect.py", line 2825, in from_callable
    follow_wrapper_chains=follow_wrapped)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/inspect.py", line 2362, in _signature_from_callable
    'no signature found for builtin type {!r}'.format(obj))
ValueError: no signature found for builtin type <class 'str'>
```

**Comment by islean** [comment]
Changed BeforeValidator(str) -> BeforeValidator(lambda x: str(x)) and now test goes well - cg upload upload-case-to-scout <case_id> constructs the models properly.

**Comment by islean** [comment]
Deployed to production 2023-08-30:

```
θ69° [isak.ohlsson@hasta:~] [P_base] $ bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-prod.sh -e P_main -t cg -b master -a

...

Already on 'master'
Your branch is behind 'origin/master' by 8 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Updating 8fb2811..44b7b0d
Fast-forward
 .bumpversion.cfg                                                                                              |  2 +-
 cg/__init__.py                                                                                                |  2 +-
 cg/apps/crunchy/models.py                                                                                     | 10 +++++-----
 cg/apps/scout/scout_export.py                                                                                 | 73 ++++++++++++++++++++++++++-----------------------------------------------
 cg/apps/scout/scoutapi.py                                                                                     |  2 +-
 cg/apps/scout/validators.py                                                                                   | 11 +++++++++++
 cg/apps/sequencing_metrics_parser/models/bcl2fastq_metrics.py                                                 | 22 ----------------------
 cg/apps/tb/api.py                                                                                             | 12 ++++++------
 cg/apps/tb/models.py                                                                                          | 72 +++++++++++++++++++++++++++++++++---------------------------------------
 cg/apps/tb/validators.py                                                                                      | 15 +++++++++++++++
 setup.py                                                                                                      |  2 +-
 tests/apps/crunchy/test_compress_fastq.py                                                                     |  2 +-
 tests/apps/scout/test_scout_models.py                                                                         | 83 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------
 tests/apps/sequencing_metrics_parser/parsers/conftest.py                                                      | 10 ----------
 tests/fixtures/apps/demultiplexing/demultiplexed-runs/170407_ST-E00198_0209_BHHKVCALXX/l1t11/Stats/Stats.json |  9 ---------
 tests/fixtures/apps/demultiplexing/demultiplexed-runs/170407_ST-E00198_0209_BHHKVCALXX/l1t21/Stats/Stats.json |  9 ---------
 tests/meta/upload/mutacc/conftest.py                                                                          |  9 ++++-----
 tests/mocks/tb_mock.py                                                                                        |  8 ++++----
 18 files changed, 179 insertions(+), 174 deletions(-)
 create mode 100644 cg/apps/scout/validators.py
 create mode 100644 cg/apps/tb/validators.py
/home/isak.ohlsson
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/isak.ohlsson
done.
Log deploy... done.
cg, version 45.0.5
```

**Pull request merged**